MINOR: add quick start for tiered storage feature

showuon opened the pull request:
Some users complained they don't have a way to determine if there is something wrong in the RSM plug-in they implemented, or there's something wrong in Kafka itself. Also, if there are users who just want to try the tiered storage feature out before implementing anything, it would be good we have an RSM implementation by default. 

Per the discussion in the KIP, there will be no default RSM implementation in Kafka, but we can use the `LocalTieredStorage` implemented for integration test, to resolve the issues above.

![image](https://github.com/apache/kafka/assets/43372967/52435eb0-28f4-448a-8a86-f4069f931472)


![image](https://github.com/apache/kafka/assets/43372967/e90ddf95-6d60-462c-bbbf-95ada1b97f6b)


![image](https://github.com/apache/kafka/assets/43372967/80f96030-a36c-4110-9b41-c6126a75db6f)



### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

showuon committed:
add quick start for tiered storage feature

showuon commented:
@kamalcph @divijvaidya @satishd , call for review.

clolov commented on docs/ops.html:
```diff
@@ -3984,27 +3984,73 @@ <h5 class="anchor-heading"><a id="tiered_storage_config_topic" class="anchor-lin
+
+<pre>
+# remote.storage.enable=true -> enabling tiered storage
+# local.retention.ms=1000 -> the time before local segments are moving to remote storage (and then deleted)
```
```suggestion
# local.retention.ms=1000 -> the time after which local copies of segments moved to remote storage are legible for deletion
```

kamalcph replied:
We can copy the doc from the TopicConfig:

```
The number of milliseconds to keep the local log segment before it gets deleted. Note that a local log segment is eligible for deletion only after it gets uploaded to remote.
```

clolov reviewed (commented):
I just have a nit about the description of the local.retention.ms. I will now proceed to try this guide out for myself!

kamalcph commented on docs/ops.html:
```diff
@@ -3984,27 +3984,73 @@ <h5 class="anchor-heading"><a id="tiered_storage_config_topic" class="anchor-lin
+</p>
+
+<p>To adopt the `LocalTieredStorage`, the test library needs to be built locally</p>
+<pre>./gradlew :storage:testJar</pre>
```
```suggestion
<pre>./gradlew clean :storage:testJar</pre>
```

kamalcph commented on docs/ops.html:
```diff
@@ -3984,27 +3984,73 @@ <h5 class="anchor-heading"><a id="tiered_storage_config_topic" class="anchor-lin
+
+<p>To adopt the `LocalTieredStorage`, the test library needs to be built locally</p>
+<pre>./gradlew :storage:testJar</pre>
+<p>After build successfully, there should be a `kafka-storage-x.x.x-SNAPSHOT-test.jar` file under `storage/build/libs`.
```
Should we ask the user to checkout the 3.6 branch and the build the jar (or) from trunk itself?

showuon replied:
Added:
```
# please checkout to the specific version tag you're using before building it
# ex: `git checkout 3.6.0`
./gradlew clean :storage:testJar
```

kamalcph commented on docs/ops.html:
```diff
@@ -3984,27 +3984,73 @@ <h5 class="anchor-heading"><a id="tiered_storage_config_topic" class="anchor-lin
+# Please provide the implementation info for remoteStorageManager. This is the mandatory configuration for tiered storage.
+# Here, we use the `LocalTieredStorage` built above.
+remote.log.storage.manager.class.name=org.apache.kafka.server.log.remote.storage.LocalTieredStorage
+remote.log.storage.manager.class.path=/Users/luke/dev/kafka/storage/build/libs/kafka-storage-x.x.x-SNAPSHOT-test.jar
```
The below configs are configured with default values but adding them to the quick-start page will add clarity on rsm and rlmm properties as they can be custom-made:
```
remote.log.storage.manager.impl.prefix=rsm.config.
remote.log.metadata.manager.impl.prefix=rlmm.config.
```

showuon replied:
Added

kamalcph commented on docs/ops.html:
```diff
@@ -3984,27 +3984,73 @@ <h5 class="anchor-heading"><a id="tiered_storage_config_topic" class="anchor-lin
+<p>After broker is started, creating a topic with tiered storage enabled with configs:
+
+<pre>
+# remote.storage.enable=true -> enabling tiered storage
```
enabling tiered storage -> enables tiered storage on the topic

kamalcph commented on docs/ops.html:
```diff
@@ -3984,27 +3984,73 @@ <h5 class="anchor-heading"><a id="tiered_storage_config_topic" class="anchor-lin
+# remote.storage.enable=true -> enabling tiered storage
+# local.retention.ms=1000 -> the time before local segments are moving to remote storage (and then deleted)
+# retention.ms=300000 -> when segments exceed this time, the segments in remote storage will be deleted
+# segment.bytes=1000000 -> for test only, to speed up the log segment rolling interval
```
Can we change the `segment.bytes` to MiB format? 

```suggestion
# segment.bytes=1048576 -> for test only, to speed up the log segment rolling interval
```

kamalcph commented on docs/ops.html:
```diff
@@ -3984,27 +3984,73 @@ <h5 class="anchor-heading"><a id="tiered_storage_config_topic" class="anchor-lin
+# local.retention.ms=1000 -> the time before local segments are moving to remote storage (and then deleted)
+# retention.ms=300000 -> when segments exceed this time, the segments in remote storage will be deleted
+# segment.bytes=1000000 -> for test only, to speed up the log segment rolling interval
+# file.delete.delay.ms=10000 -> for test only, to speed up the file delete delay
```
```suggestion
# file.delete.delay.ms=10000 -> for test only, to speed up the local-log segment file delete delay
```

kamalcph commented on docs/ops.html:
```diff
@@ -3984,27 +3984,73 @@ <h5 class="anchor-heading"><a id="tiered_storage_config_topic" class="anchor-lin
+<pre>
+# remote.storage.enable=true -> enabling tiered storage
+# local.retention.ms=1000 -> the time before local segments are moving to remote storage (and then deleted)
+# retention.ms=300000 -> when segments exceed this time, the segments in remote storage will be deleted
```
Can we increase the `retention.ms` to 1 hour? The default is 7 days. User might gets confused why they are not seeing all the messages that were produced within 5 mins.

kamalcph commented on docs/ops.html:
```diff
@@ -3984,27 +3984,73 @@ <h5 class="anchor-heading"><a id="tiered_storage_config_topic" class="anchor-lin
+
+<p>Last, we can try to consume some data in the offset 0, to make sure it will successfully fetch from the remote storage.</p>
+
+<pre>bin/kafka-console-consumer.sh --topic tieredTopic --partition 0 --offset 0 --max-messages 1 --bootstrap-server localhost:9092</pre>
```
```suggestion
<pre>bin/kafka-console-consumer.sh --topic tieredTopic --partition 0 --from-beginning --max-messages 1 --bootstrap-server localhost:9092 --property print.offset=true</pre>
```

kamalcph commented on docs/ops.html:
```diff
@@ -3984,27 +3984,73 @@ <h5 class="anchor-heading"><a id="tiered_storage_config_topic" class="anchor-lin
+# Configure the directory used for `LocalTieredStorage`
+rsm.config.dir=/tmp/kafka-remote-storage
+
+# This needs to be changed if broker numbers are more than 1
```
```suggestion
# This needs to be changed if number of brokers in the cluster is more than 1
```

kamalcph commented on docs/ops.html:
```diff
@@ -3984,27 +3984,73 @@ <h5 class="anchor-heading"><a id="tiered_storage_config_topic" class="anchor-lin
+<p>Try to send messages to the `tieredTopic` topic to roll the log segment:</p>
+
+<pre>
+bin/kafka-producer-perf-test.sh --topic tieredTopic --num-records 1000 --record-size 1024 --throughput -1 --producer-props bootstrap.servers=localhost:9092
```
If we are changing the `segment.bytes` to MiB format, then the record-size might be tuned slightly to ensure that the segment gets rotated.

showuon replied:
Updated the num-records to 1200 and tested locally.

kamalcph reviewed (commented):
Left minor comments to address. Thanks for adding the tiered storage quick-start!

divijvaidya reviewed (commented):
1. Should we add a section on how to reset the state of system after the completion of quickstart? This is important because folks trying this out may not know what side-effects does running this quick start has. As an example, we should add a section to delete the tiered storage directory to reset the state.

2. At the places where we say "start the broker", can we add a hyperlink to the other quickstart that details about how to start a kafka broker? Asking this because as of now, this quick start is not self sufficient, i.e. a reader will have to switch to other places to complete this quickstart. Personally, I find quickstart guides to be very helpful if they are self-sufficient i.e. I keep on simply copy and pasting commands, and it just works.


Of course, we can do both of the above incrementally after this PR. This is a great start!

clolov commented:
Apologies for the delay, I ran the commands in the guide and they work! My only question is whether it is worthwhile directly adding the necessary configurations to the server.properties which comes with Kafka and commenting out the ones which will activate the sub-system? One way or another people will have to add those configurations and this might save them a bit of effort?

satishd commented on docs/ops.html:
```diff
@@ -3984,27 +3984,73 @@ <h5 class="anchor-heading"><a id="tiered_storage_config_topic" class="anchor-lin
+<h4 class="anchor-heading"><a id="tiered_storage_config_ex" class="anchor-link"></a><a href="#tiered_storage_config_ex">Quick Start Example</a></h4>
+
+<p>Apache Kafka doesn't provide an out-of-the-box RemoteStorageManager implementation. To have a preview of the tiered storage
+  feature, the `LocalTieredStorage` implemented for integration test can be used, which will create a temporary directory in local storage
```
Good to give hyper link for `LocalTieredStorage`.

showuon replied:
Agree. Added.

satishd commented on docs/ops.html:
```diff
@@ -3984,27 +3984,73 @@ <h5 class="anchor-heading"><a id="tiered_storage_config_topic" class="anchor-lin
+remote.log.storage.manager.class.name=org.apache.kafka.server.log.remote.storage.LocalTieredStorage
+remote.log.storage.manager.class.path=/Users/luke/dev/kafka/storage/build/libs/kafka-storage-x.x.x-SNAPSHOT-test.jar
+
+# Configure the directory used for `LocalTieredStorage`
```
Should we also clarify that all the brokers need to have access to this directory with the same path configured here?

showuon replied:
Added a note:
```
# Configure the directory used for `LocalTieredStorage`
# Note, please make sure the brokers need to have access to this directory
rsm.config.dir=/tmp/kafka-remote-storage
```

satishd reviewed (commented):
Thanks @showuon for started working on improving the existing Example docs. It is important to have a nice quick start guide for users to try out this feature.

showuon committed:
Merge branch 'trunk' of https://github.com/apache/kafka into quickStartTiered

showuon committed:
refactor quick strat tiered storage

showuon commented:
@clolov @divijvaidya @satishd @kamalcph , I've addressed all your comments and updated the screenshot in the PR description. Please take a look again. Thanks.

divijvaidya reviewed (approved):
We can keep on improving this iteratively but this is great start! LGTM.

kamalcph reviewed (approved):
LGTM

satishd reviewed (approved):
Thanks @showuon for addressing the review comments. LGTM.

showuon commented:
> whether it is worthwhile directly adding the necessary configurations to the server.properties which comes with Kafka and commenting out the ones which will activate the sub-system? One way or another people will have to add those configurations and this might save them a bit of effort?

@clolov , I'll work on that in a separate PR. Let's update the docs soon to let users try the tiered storage feature first.
